runtime/common/src/main/kotlin/kotlinx/serialization/json/internal/Polymorphic.kt

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2017-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization.json.internal
+
+import kotlinx.serialization.*
+import kotlinx.serialization.internal.*
+import kotlinx.serialization.json.*
+
+internal inline fun <T> JsonOutput.encodePolymorphically(serializer: SerializationStrategy<T>, value: T, ifPolymorphic: () -> Unit) {
+    if (serializer !is PolymorphicSerializer<*> || json.configuration.useArrayPolymorphism) {
+        serializer.serialize(this, value)
+        return
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    val actualSerializer = serializer.findPolymorphicSerializer(this, value as Any) as KSerializer<Any>
+    if (actualSerializer is EnumSerializer<*>) {
```

Comment:
Should there be the same check for primitives? Also, any serializer which would represent classes as value, not as a composite seems to be in trouble here. Probably we can make use of checking a `descriptor.kind`